api/pkg/transformer/server/grpc/server.go

```diff
@@ -0,0 +1,249 @@
+	"github.com/gojek/merlin/pkg/transformer/server/instrumentation"
+	"github.com/gojek/merlin/pkg/transformer/types"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+
+	"github.com/afex/hystrix-go/hystrix"
+	upiv1 "github.com/caraml-dev/universal-prediction-interface/gen/go/grpc/caraml/upi/v1"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/reflection"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	shutdownSignals = []os.Signal{os.Interrupt, syscall.SIGTERM}
+)
+
+// UPIServer serves GRPC request that implement UPI Service
+type UPIServer struct {
+	upiv1.UnimplementedUniversalPredictionServiceServer
+
+	opts        *config.Options
+	modelClient upiv1.UniversalPredictionServiceClient
+	conn        *grpc.ClientConn
+	logger      *zap.Logger
+
+	// ContextModifier function to modify or store value in a context
+	ContextModifier func(ctx context.Context) context.Context
+	// PreprocessHandler function to run all preprocess operation
+	PreprocessHandler func(ctx context.Context, request types.Payload, requestHeaders map[string]string) (types.Payload, error)
+	// PostprocessHandler function to run all preprocess operation
+	PostprocessHandler func(ctx context.Context, response types.Payload, responseHeaders map[string]string) (types.Payload, error)
+}
+
+// NewUPIServer creates GRPC server that implement UPI Service
+func NewUPIServer(opts *config.Options, handler *pipeline.Handler, logger *zap.Logger) (*UPIServer, error) {
+	dialOpts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	}
+	conn, err := grpc.Dial(opts.ModelPredictURL, dialOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	hystrix.ConfigureCommand(opts.ModelGRPCHystrixCommandName, hystrix.CommandConfig{
+		Timeout:                hystrixpkg.DurationToInt(opts.ModelTimeout, time.Millisecond),
+		MaxConcurrentRequests:  opts.ModelHystrixMaxConcurrentRequests,
+		RequestVolumeThreshold: opts.ModelHystrixRequestVolumeThreshold,
+		SleepWindow:            opts.ModelHystrixSleepWindowMs,
+		ErrorPercentThreshold:  opts.ModelHystrixErrorPercentageThreshold,
+	})
+
+	modelClient := upiv1.NewUniversalPredictionServiceClient(conn)
+	svr := &UPIServer{
+		opts:        opts,
+		modelClient: modelClient,
+		conn:        conn,
+		logger:      logger,
+	}
+
+	if handler != nil {
+		svr.ContextModifier = handler.EmbedEnvironment
+		svr.PreprocessHandler = handler.Preprocess
+		svr.PostprocessHandler = handler.Postprocess
+	}
+
+	return svr, nil
+}
+
+// Run running GRPC Server
+func (us *UPIServer) Run() {
+	// defer us.conn.Close()
+	lis, err := net.Listen("tcp", fmt.Sprintf("localhost:%s", us.opts.GRPCPort))
+	if err != nil {
+		us.logger.Error(fmt.Sprintf("failed to listen the port %s", us.opts.GRPCPort))
+		return
+	}
+	opts := []grpc.ServerOption{
```

Comment:
What are commonly options that need to be configurable? can we create a card/task to make it configurable? Same goes for the client config in line 56-59

Reply:
sure, I think we can work on the configurable options after do some load test